src/stateManager/StateManager.ts

```diff
@@ -278,122 +273,80 @@ class StateManager {
-                    this.storeExitChannelBlock(
-                        applyResult.exitChannels,
-                        block.coordinates
-                    );
-
-                    this.handleStateSnapshotStorage(
-                        applyResult.encodedState,
-                        block.forkId
-                    );
-
-                    this.storage.blocks.storeBlock(signedBlock);
-
-                    // Schedule the success callback
-                    scheduleTask(
-                        applyResult.successCallback,
-                        0,
-                        "stateTransitionSuccessCallback"
-                    );
-                }
-            }
-
-            return finalExecutionFlag;
-        } finally {
-            // Safety check: must have an execution flag
-            if (finalExecutionFlag === undefined) {
-                throw new Error(
-                    "StateManager - onSignedBlock - Internal Error - flag undefined"
-                );
-            }
-
-            // Process the final decision
-            await this.executionDecisionProcessor.process(
-                signedBlock,
-                finalExecutionFlag,
-                finalAgreementFlag
-            );
-            this.mutex.unlock();
-        }
+        return this.onBlockConfirmation({
+            signedBlock,
+            signatures: []
+        });
     }
 
     // Passes the block confirmation through a verification pipeline and returns an execution flag
     public async onBlockConfirmation(
-        signedBlock: SignedBlockStruct,
-        confirmationSignature: Signature,
-        block?: Block
+        blockConfirmation: BlockConfirmationStruct
     ): Promise<ExecutionFlags> {
-        let finalExecutionFlag: ExecutionFlags = ExecutionFlags.SUCCESS; // Default to SUCCESS
-        const decodedBlock = block ?? Block.decode(signedBlock.encodedBlock);
+        const validationFlag =
+            await this.validationService.validateBlockConfirmation(
+                blockConfirmation
+            );
 
+        // If validation failed or is not ready, return early
+        if (validationFlag !== ExecutionFlags.SUCCESS) {
+            return validationFlag;
+        }
+        const block = Block.decode(blockConfirmation.signedBlock.encodedBlock);
+
+        // apply state transition and validate state transition result
         try {
-            const result =
-                await this.validationService.validateBlockConfirmation(
-                    signedBlock,
-                    confirmationSignature,
-                    decodedBlock
-                );
-            finalExecutionFlag = result.flag;
+            const {
+                success,
+                encodedState,
+                previousStateHash,
+                successCallback,
+                exitChannels
+            } = await this.applyTransaction(block.transaction);
```

Comment:
in general since `BlockConfirmation pipeline` is the largest part of the `StateManager` you extract and implement that pipeline in a seperate file as part of `StateManager` e.g. `BlockConfirmationPipeline.ts` instead of `ValidationService.ts` - stateTransition stuff can also be a seperate file so you can re-use it there and for `playTransaction`

That will also help with not decoding the block twice as on line 298 (a few above)

btw `successCallback` and `exitChannels` are not used - they should be part of `success()`

Reply:
or `BlockConfirmationService` if we're using that convention